### PR TITLE
refactor: remove sessions

### LINKED ISSUE #394: Ditch sessions from activities
This flow should be rethought after the 1.0.0 release.

### PR DESCRIPTION
**Pull request by danielsp45**

### CODE REVIEW

**Review by ruilopesm — CHANGES_REQUESTED** [review]
There are a few pages left.
- Calendar (week and month components too)
- Edit Activity
- Show Organization
- Show Department

File: lib/atomic/activities/activity.ex [hunk 1]
```diff
@@ -4,21 +4,40 @@ defmodule Atomic.Activities.Activity do
...
-      on_replace: :delete_if_exists,
-      foreign_key: :activity_id,
-      preload_order: [asc: :start]
+    field :start, :naive_datetime
+    field :finish, :naive_datetime
+    field :image, Uploaders.Post.Type
+    field :maximum_entries, :integer
+    field :minimum_entries, :integer
+    field :enrolled, :integer, virtual: true
+    embeds_one :location, Location, on_replace: :delete
```
Comment (ruilopesm) [inline lib/atomic/activities/activity.ex]:
```suggestion
    embeds_one :location, Location
```

File: lib/atomic/activities/activity.ex [hunk 2]
```diff
@@ -4,21 +4,40 @@ defmodule Atomic.Activities.Activity do
...
+
+    many_to_many :speakers, Speaker, join_through: ActivitySpeaker, on_replace: :delete
```
Comment (ruilopesm) [inline lib/atomic/activities/activity.ex]:
```suggestion
    many_to_many :speakers, Speaker, join_through: ActivitySpeaker
```

File: lib/atomic/activities/activity.ex [hunk 3]
```diff
@@ -4,21 +4,40 @@ defmodule Atomic.Activities.Activity do
...
+    many_to_many :departments, Department, join_through: ActivityDepartment, on_replace: :delete
```
Comment (ruilopesm) [inline lib/atomic/activities/activity.ex]:
```suggestion
    many_to_many :departments, Department, join_through: ActivityDepartment
```

File: lib/atomic/activities/activity.ex [hunk 4]
```diff
@@ -28,7 +47,74 @@ defmodule Atomic.Activities.Activity do
...
+      Ecto.Changeset.put_assoc(changeset, :departments, departments)
+    else
+      changeset
+    end
+  end
+
+  defp maybe_put_speakers(changeset, attrs) do
+    if attrs["speakers"] do
+      speakers = Activities.get_speakers(attrs["speakers"])
+
```
Comment (ruilopesm) [inline lib/atomic/activities/activity.ex:113]:
```suggestion
```

File: lib/atomic/activities/activity.ex [hunk 5]
```diff
@@ -28,7 +47,74 @@ defmodule Atomic.Activities.Activity do
...
+      %{changeset | action: :delete}
+    else
+      changeset
+    end
+  end
+
+  defp maybe_put_departments(changeset, attrs) do
+    if attrs["departments"] do
+      departments = Departments.get_departments(attrs["departments"])
+
```
Comment (ruilopesm) [inline lib/atomic/activities/activity.ex]:
```suggestion
```

File: lib/atomic/activities/speaker.ex [hunk 6]
```diff
@@ -14,7 +14,7 @@ defmodule Atomic.Activities.Speaker do
     field :name, :string
     field :bio, :string
 
-    many_to_many :activities, Activity, join_through: SessionSpeaker, on_replace: :delete
+    many_to_many :activities, Activity, join_through: ActivitySpeaker, on_replace: :delete
```
Comment (ruilopesm) [inline lib/atomic/activities/speaker.ex]:
```suggestion
    many_to_many :activities, Activity, join_through: ActivitySpeaker
```

File: lib/atomic/departments.ex [hunk 7]
```diff
@@ -52,18 +52,18 @@ defmodule Atomic.Departments do
...
   """
-  def get_department_sessions(department_id) do
-    Repo.all(from s in SessionDepartment, where: s.department_id == ^department_id)
-    |> Repo.preload([:session, session: :activity])
-    |> Enum.map(& &1.session)
+  def get_department_activities(department_id) do
+    Repo.all(from a in ActivityDepartment, where: a.department_id == ^department_id)
+    |> Repo.preload([:activity])
+    |> Enum.map(& &1.activity)
   end
```
Comment (ruilopesm) [inline lib/atomic/departments.ex]:
```suggestion
  def get_department_activities(department_id, opts \\ [preloads: [:activity]]) do
    ActivityDepartment
    |> where([a], a.department_id == ^department_id)
    |> apply_filters(opts)
    |> select([a], a.activity)
    |> Repo.all()
  end
```
I think this should work, but, please, ensure it.
(Maybe the `select` and the `apply_filters` calls should be inverted)

File: lib/atomic/organizations/department.ex [hunk 8]
```diff
@@ -13,7 +14,7 @@ defmodule Atomic.Organizations.Department do
     field :name, :string
     field :description, :string
 
-    many_to_many :sessions, Session, join_through: SessionDepartment, on_replace: :delete
+    many_to_many :activities, Activity, join_through: ActivityDepartment, on_replace: :delete
```
Comment (ruilopesm) [inline lib/atomic/organizations/department.ex]:
```suggestion
    many_to_many :activities, Activity, join_through: ActivityDepartment
```

File: lib/atomic/departments.ex (hunk 7 above)
Comment (danielsp45) [reply lib/atomic/departments.ex]: I think it maybe should be after, so it applies all the filters in a reduced number of entries.

**Review by ruilopesm — APPROVED** [review]

**Pull request merged**